File: Analyzer/python/RaceMinerLib/Manager/DBMgr/LockCollectionMgr.py
import logging
from dataclasses import dataclass
from typing import List

from RaceMinerLib.Manager.DBMgr import DBMgr
from RaceMinerLib.Manager.DBMgr import TBLMgr
from RaceMinerLib.Manager.PathMgr import PathMgr


@dataclass
class LockCollectionRow:
    file: str
    func: str
    lock_func: str
    lock_var: str
    lock_line: int
    unlock_func: str
    unlock_var: str
    unlock_line: int
    access_var: str
    access_line: int
    access_type: str


class LockCollectionMgr(TBLMgr):
    def __init__(self, path_mgr, db_mgr):
        super().__init__(db_mgr)

        self.path_mgr = path_mgr
        self.db_mgr = db_mgr

        self.logger = logging.getLogger(self.__class__.__name__)
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                "[%(asctime)s][%(levelname)s] %(message)s"
            )
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)
            self.logger.setLevel(logging.INFO)

        sql_path = self.path_mgr.get_tbl_create_lock_collection()
        sql = self.load_sql(sql_path)

        if sql:
            self.db_mgr.exec_simple(sql)
            self.logger.info("LockCollection table initialized")
        else:
            self.logger.error("Failed to initialize LockCollection table")

    def insert(self, row: LockCollectionRow) -> bool:
        sql_path = self.path_mgr.get_tbl_insert_lock_collection()
        sql = self.load_sql(sql_path)

        if not sql:
            return False

        params = [
            row.file, row.func,
            row.lock_func, row.lock_var, row.lock_line,
            row.unlock_func, row.unlock_var, row.unlock_line,
            row.access_var, row.access_line, row.access_type
        ]

        return self.db_mgr.exec_sql(sql, params)

    def insert_batch(self, row_list: List[LockCollectionRow]) -> bool:
        if not row_list:
            return True

        sql_path = self.path_mgr.get_tbl_insert_lock_collection()
        sql = self.load_sql(sql_path)

        if not sql:
            return False

        params_list = [
            [
                r.file, r.func,
                r.lock_func, r.lock_var, r.lock_line,
                r.unlock_func, r.unlock_var, r.unlock_line,
                r.access_var, r.access_line, r.access_type
            ]
            for r in row_list
        ]

        return self.db_mgr.exec_sql_batch(sql, params_list)

    def select_all(self) -> List[LockCollectionRow]:
        sql_path = self.path_mgr.get_tbl_select_lock_collection()
        sql = self.load_sql(sql_path)

        results: List[LockCollectionRow] = []

        if not sql:
            return results

        cur = self.db_mgr.prepare_query(sql, [])
        if not cur:
            return results

        try:
            rows = cur.fetchall()

            for r in rows:
                row = LockCollectionRow(
                    file=r[1],
                    func=r[2],
                    lock_func=r[3],
                    lock_var=r[4],
                    lock_line=r[5],
                    unlock_func=r[6],
                    unlock_var=r[7],
                    unlock_line=r[8],
                    access_var=r[9],
                    access_line=r[10],
                    access_type=r[11],
                )
                results.append(row)

        except Exception as e:
            self.logger.error(f"[select_all] {e}")

        finally:
            self.db_mgr.finalize_query(cur)

        return results

File: Analyzer/cpp/lib/Manager/DBMgr/LockCollectionMgr.cpp
#include "Manager/DBMgr/LockCollectionMgr.h"

#include "Utils/DBOperations.h"

using namespace std;

/*
 * Implementation of LockCollectionMgr
 */

LockCollectionMgr::LockCollectionMgr(PathMgr *path_mgr, DBMgr *db_mgr) : TBLMgr(db_mgr) {
	this->path_mgr = path_mgr;
	this->db_mgr = db_mgr;
	string tbl_create_lock_collection = path_mgr->getTblCreateLockCollection();
	string tbl_create_lock_collection_sql = loadSql(tbl_create_lock_collection);
	db_mgr->execSimple(tbl_create_lock_collection_sql);
}

LockCollectionMgr::~LockCollectionMgr() {
	// Need to do nothing at present.
}

bool LockCollectionMgr::insert(LockCollectionRow &row) {
	SqlParams sql_params = {row.file, row.func, 
					row.lock_func, row.lock_var, row.lock_line,
					row.unlock_func, row.unlock_var, row.unlock_line,
					row.access_var, row.access_line, row.access_type};
	string tbl_insert_lock_collection = path_mgr->getTblInsertLockCollection();
	string tbl_insert_lock_collection_sql = loadSql(tbl_insert_lock_collection);
	return db_mgr->execSql(tbl_insert_lock_collection_sql, sql_params);
}

bool LockCollectionMgr::insertBatch(vector<LockCollectionRow> &row_vec) {
	vector<SqlParams> params_list;
	params_list.reserve(row_vec.size());

	for (size_t row_idx = 0; row_idx < row_vec.size(); row_idx++) {
		LockCollectionRow &row = row_vec[row_idx];
		params_list.push_back({row.file, row.func,
					row.lock_func, row.lock_var, row.lock_line,
					row.unlock_func, row.unlock_var, row.unlock_line,
					row.access_var, row.access_line, row.access_type});
	}
	string tbl_insert_lock_collection = path_mgr->getTblInsertLockCollection();
	string tbl_insert_lock_collection_sql = loadSql(tbl_insert_lock_collection);
	return db_mgr->execSqlBatch(tbl_insert_lock_collection_sql, params_list);
}

vector<LockCollectionRow> LockCollectionMgr::selectAll() {
	string tbl_select_lock_collection = path_mgr->getTblSelectLockCollection();
	string tbl_select_lock_collection_sql = loadSql(tbl_select_lock_collection);
	vector<LockCollectionRow> results;

	sqlite3_stmt * stmt = db_mgr->prepareQuery(tbl_select_lock_collection_sql, {});
	if (!stmt) {
		return results;
	}
	while (sqlite3_step(stmt) == SQLITE_ROW) {
		LockCollectionRow row;
		row.file = get_column_text(stmt, 1);
		row.func = get_column_text(stmt, 2);
		row.lock_func = get_column_text(stmt, 3);
		row.lock_var = get_column_text(stmt, 4);
		row.lock_line = get_column_int(stmt, 5);
		row.unlock_func = get_column_text(stmt, 6);
		row.unlock_var = get_column_text(stmt, 7);
		row.unlock_line = get_column_int(stmt, 8);
		row.access_var = get_column_text(stmt, 9);;
		row.access_line = get_column_int(stmt, 10);
		row.access_type = get_column_text(stmt, 11);

		results.push_back(row);
	}

	db_mgr->finalizeQuery(stmt);
	return results;
}

vector<LockCollectionRow> LockCollectionMgr::selectFile(string src_file) {
	string tbl_select_file_lock_collection = path_mgr->getTblSelectFileLockCollection();
	string tbl_select_file_lock_collection_sql = loadSql(tbl_select_file_lock_collection);
	vector<LockCollectionRow> results;

	sqlite3_stmt * stmt = db_mgr->prepareQuery(tbl_select_file_lock_collection_sql, {src_file});
	if (!stmt) {
		return results;
	}
	while (sqlite3_step(stmt) == SQLITE_ROW) {
		LockCollectionRow row;
		row.file = get_column_text(stmt, 1);
		row.func = get_column_text(stmt, 2);
		row.lock_func = get_column_text(stmt, 3);
		row.lock_var = get_column_text(stmt, 4);
		row.lock_line = get_column_int(stmt, 5);
		row.unlock_func = get_column_text(stmt, 6);
		row.unlock_var = get_column_text(stmt, 7);
		row.unlock_line = get_column_int(stmt, 8);
		row.access_var = get_column_text(stmt, 9);;
		row.access_line = get_column_int(stmt, 10);
		row.access_type = get_column_text(stmt, 11);

		results.push_back(row);
	}

	db_mgr->finalizeQuery(stmt);
	return results;
}
